### ISSUE TITLE
[BUG] Aggregation query with custom check permission does not work

### ISSUE THREAD

**Issue by munjalpatel** [hasura/graphql-engine#7660]
### Version Information

Server Version: `2.0.7`

### Environment

Self-hosted

### What is the expected behavior?

Aggregation queries with "custom checks" should not return count for records that do not satisfy the rules in the "custom checks."

### Keywords

aggregate permissions, aggregation permissions

### What is the current behavior?

Aggregate queries do not respect "custom checks" in the permissions and returns `count` of all records.

### How to reproduce the issue?

1. Create these 3 tables:
```
content (id uuid, data text, source_id uuid)
sources (id uuid, name text, collection_id uuid)
collections (id uuid, name text, team_id uuid)
```

2. Create the following relationships:
```
content.source_id --> sources.id
sources.collection_id --> collections.id
```

3. Add the following permission on the `content` table:

![image](https://user-images.githubusercontent.com/2717686/135168485-37222a09-8e14-4b97-a9e4-778e1ceb848a.png)

4. Add a few records in all tables with different `team_Id`

5. Run the following query with the `role` you added permission for.
```graphql
query {
  content_aggregate {
    aggregate {
      count
    }
  }
}
```

Notice how count for all records is returned instead of records for a specific `team_id`

** `team_id` value is properly passed to Hasura in JWT and is being used in many other queries and permissions without any issues.

**Comment by munjalpatel** [comment]
@tirumaraiselvan @BenoitRanque I verified that the `x-hasura-admin-secret` header is not passed.

Here is what Hasura logs once I make a request:
```
{
  "type": "http-log",
  "timestamp": "2021-10-08T03:23:27.868+0000",
  "level": "info",
  "detail": {
    "operation": {
      "query_execution_time": 0.104745381,
      "user_vars": {
        "x-hasura-role": "owner",
        "x-hasura-plan": "premium",
        "x-hasura-team-id": "7709aef7-36dd-46b5-91cd-97d3a31e0edd",
        "x-hasura-user-id": "79d8c016-96a9-4b37-94a1-4d9f505402f0"
      },
      "request_id": "92dbf125-2de6-9481-b2ee-2f98bf47bf3e",
      "response_size": 99,
      "request_mode": "single",
      "request_read_time": 7.462e-6
    },
    "request_id": "92dbf125-2de6-9481-b2ee-2f98bf47bf3e",
    "http_info": {
      "status": 200,
      "http_version": "HTTP/1.1",
      "url": "/v1/graphql",
      "ip": "10.0.107.254",
      "method": "POST",
      "content_encoding": "gzip"
    }
  }
}
```

**Comment by munjalpatel** [comment]
I have not selected any "column" permissions for this one, would that cause issues like this?

**Comment by plcplc** [comment]
Hm. @BenoitRanque, your reproduction case has:

> Headers:
> 
> ```
> x-hasura-admin-secret <secret>
> ```

.. But previously you questioned whether `x-hasura-admin-secret` might have been supplied in error:

> Is it possible the user accidentally submitted the `x-hasura-admin-secret` header?
> When that header is submitted, the `Authorization` header is ignored, so you would see this kind of result.

Is this maybe the reason you were able to reproduce the issue or did it turn out that the admin secret had no influence after all?

**Comment by BenoitRanque** [comment]
@plcplc 

> Is this maybe the reason you were able to reproduce the issue or did it turn out that the admin secret had no influence after all?

The `x-hasura-admin-secret` does mean that the `Authorization` header will be skipped, but it also means that the `x-hasura-role` header will be respected (instead of being validated against the claims in the token in the `Authorization` header)

**Comment by eviefp** [comment]
@BenoitRanque Just to be sure, the expected result for this query (given your repro setup)

```graphql
query MyQuery {
  car_aggregate {
    aggregate {
      count
    }
  }
  car {
    name
    owner_id
  }
}
```

would be

```json
{
  "data": {
    "car_aggregate": {
      "aggregate": {
        "count": 2
      }
    },
    "car": [
      {
        "name": "red",
        "owner_id": 1
      },
      {
        "name": "blue",
        "owner_id": 1
      },
      {
        "name": "green",
        "owner_id": 2
      }
    ]
  }
}
```

Which is to say, when aggregating, we should respect the aggregation filter, even if the inherited role allows us to select more data than the aggregation. Correct?

**Comment by BenoitRanque** [comment]
@eviefp yes, this is correct. In our example, the user has access to select all records, but only has aggregate permissions on a subset of records. We would expect the result you are showing.

**Comment by tirumaraiselvan** [comment]
The spec says:

1. The inherited role has permission to access all rows by ORing the filters of individual rows. In the referenced [example](https://github.com/hasura/graphql-engine/issues/7660#issuecomment-939098742), that means all rows.

2. The allow aggregate permission will be allowed if any of the roles allows it.

Hence, the expected result is correct for the aggregate. We will keep the issue open as feedback for further customization of the behaviour of inherited roles.

**Comment by plcplc** [comment]
@tirumaraiselvan : Sorry, but is it not the other way around? The expectation that `count: 2` is _not supported_ by the docs, which treat each of the permission aspects (i.e. limit, aggregate) as independent concepts.

I suspect that this is actually what you were trying to say, but I understand your comment to imply the opposite, i.e. that the expectation of `count: 2` is the correct interpretation and that the implementation thus has a bug for actually outputting `count: 3`.

For another example, the _limit_ that results from an inherited role is the maximum (i.e. a single number) of the limit of each of the roles that make up the inherited role. i.e. there is no coupling between column/row permissions and limit, and the same goes for the permission to aggregate.

Suppose you have an inherited role comprised of roles `R: {limit: 1, columns: ["A"]}` and `S: {limit: 100, columns: ["B"]}`. This results in the merged permissions `inherit(R,S): {limit: 100, columns: ["A",B"]}`. Thus the inherited role allows you to select 100 rows of column A, because there is no coupling between limit and columns. The same logic holds for the permission to aggregate.

**Comment by plcplc** [comment]
(Interestingly, an exception to this is that row and column permissions do have some dependent interaction when combined via inherited roles)

**Comment by tirumaraiselvan** [comment]
> @tirumaraiselvan : Sorry, but is it not the other way around? The expectation that count: 2 is not supported by the docs, which treat each of the permission concepts (i.e. limit, aggregate, columns, rows) as independent from each other.

@plcplc Sorry for the ambiguity in my last message. I have updated it.
